### ISSUE TITLE
Best practice for input debounce (no-noncallable-event-binding)

### ISSUE THREAD

**Issue by abdonrd** [runem/lit-analyzer#66]
Today I see this code:

<img width="883" alt="Screenshot 2019-11-12 at 19 20 38" src="https://user-images.githubusercontent.com/1007051/68698603-92cc3d00-0581-11ea-9927-6e7624be941e.png">

And the [`debounce`](https://lodash.com/docs/4.17.15#debounce) function is from `lodash-es`:

```javascript
import { debounce } from 'lodash-es';
```

The code works, but `lit-analyzer` complains.

**Comment by 43081j** [comment]
I imagine this is because the comparison will check that _all_ types are of a "function kind" and its not detecting that `Cancelable` is.

**Comment by abdonrd** [comment]
And there are any workaround like `eslint-disable-line` to ignore this case?

**Comment by runem** [comment]
@abdonrd I apologise for this bug, I'll look into it now :blush: 

In the meantime you can work around this problem in three ways:
1. In a Typescript file you can cast it to Function: `as Function`
2. You can add `<!-- @ts-ignore -->` right before the tag name. This will, however, disable all inspections for the node.
3. You can temporarily disable the "noncallable-event-binding" rule in vscode settings.
4. If you are using the CLI you can add the option `--rules.noncallable-event-binding off`

Like @43081j, I also think this is because `ts-simple-type` is not detecting that Cancelable is callable. Initially, however, when looking into the typings for `lodash-es`, it seems like the type of `Callable` is a "non-calllable" interface:

```
interface Cancelable {
    cancel(): void;
    flush(): void;
}
```

The problem is definitely a bug in my code, because Typescript thinks `Function & Callable` is in fact assignable to the `Function`. I'm also thinking that it could be a bug with the `&` operator in `ts-simple-type`. I'll have to look into what the TypeChecker thinks and set up some tests in `ts-simple-type`. I'll update you here :-)

**Comment by abdonrd** [comment]
Thanks @runem! 👏

**Comment by abdonrd** [comment]
Any update with this @runem? Thanks!

**Comment by abdonrd** [comment]
Any update with this @runem? Thanks!

**Comment by runem** [comment]
Finally I managed to refactor `ts-simple-type` to support this. I tested it on https://github.com/runem/lit-analyzer/pull/112, and `debounce` lodash function is type checked correctly now :-)

If you would like to try it out for yourself you can install the `.vsix` file from https://github.com/runem/lit-analyzer/suites/887960051/artifacts/10512866

I'm sorry it took me so long to fix this issue!

**Comment by abdonrd** [comment]
Great! I can't wait for the new version to come out!

**Comment by abdonrd** [comment]
Hi @runem!

I just check this and I have the same issue using the latest version of the VSCode plugin (v1.2.1):

<img width="1040" alt="Screenshot 2020-08-05 at 16 54 38" src="https://user-images.githubusercontent.com/1007051/89428229-73db8f80-d73c-11ea-9d62-570c5d3c88fd.png">

And the same the latest version of `lit-analyzer` (v1.2.1):

<img width="801" alt="Screenshot 2020-08-05 at 16 56 50" src="https://user-images.githubusercontent.com/1007051/89428555-cae16480-d73c-11ea-95c5-cbbcea9eed97.png">

**Comment by abdonrd** [comment]
Friendly pint to @runem. Thanks!

**Comment by PaulMaly** [comment]
Have the same issue using `1.2.4` version of VS code plugin with latest `lit-html`. Seems `lit-html` directives simply has changed its definition - `DirectiveResult `, returning now by user-facing `directive` function, is definitely not callable thing.

<img width="815" alt="Screenshot 2022-09-13 at 21 33 28" src="https://user-images.githubusercontent.com/4378873/189982664-c6005f87-05d4-43a6-803d-a3b5db97413b.png">

**Comment by 43081j** [comment]
im guessing this needs to support lit 2.x directives:
https://github.com/runem/lit-analyzer/blob/3218622f7f7212ab003bca39af6e58387c0a6d42/packages/lit-analyzer/src/lib/rules/util/type/extract-binding-types.ts#L46-L51

im not sure ill have time this week but ill try take a look
